File: 350. Intersection of Two Arrays II.cpp
class Solution {
public:
    // hashmap: O(m+n)time & space, 57.37%, ok
    vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {
        vector<int> ret;
        unordered_map<int, int> map;
        if(nums1.empty() || nums2.empty()) return ret;
        for(auto n: nums1) map[n]++;
        for(auto n: nums2){
            if(map.count(n) && map[n] > 0){
                map[n]--;
                ret.push_back(n);
            }
        }
        return ret;
    }
    
    // sort and two pointers, O(max(mlogm, nlogn)) time, O(1)space, 18.12%
    vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {
        vector<int> ret;
        if(nums1.empty() || nums2.empty()) return ret;
        sort(nums1.begin(), nums1.end());
        sort(nums2.begin(), nums2.end());
        int i = 0, j = 0;
        while(i < nums1.size() && j < nums2.size()){
            if(nums1[i] == nums2[j]){
                ret.push_back(nums1[i]);
                i++; j++;
            }
            else if(nums1[i] > nums2[j])
                j++;
            else i++;
        }
        return ret;
    }
};

File: 350. Intersection of Two Arrays II.py
class Solution(object):
    
    # hash table: O(m + n) time, O(max(m,n)) space
    def intersect(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: List[int]
        """
        count, ret = {}, []
        for num in nums1:
            count[num] = count.get(num, 0) + 1
        for num in nums2:
            if num in count and count[num] > 0:
                ret.append(num)
                count[num] -= 1
        return ret
    
    # two pointers: O(max(m,n)*log(max(m,n))) time, O(1) space
    def intersect(self, nums1, nums2):
        nums1.sort()
        nums2.sort()
        ret, i, j = [], 0, 0
        while i < len(nums1) and j < len(nums2):
            if nums1[i] == nums2[j]:
                ret.append(nums1[i])
                i, j = i + 1, j + 1
            elif nums1[i] > nums2[j]:
                j += 1
            else:
                i += 1
        return ret
            
        